XMonad/Layout/IndependentScreens.hs

```diff
@@ -186,3 +218,86 @@ unmarshallWindowSpace :: WindowSpace -> WindowSpace
+    ppSort              = fmap (marshallSort' s) (ppSort pp)
+    }
+
+-- if use original marshallSort, getSortByIndex will not work properly, since it sends the "video" instead of "0_video" back, and getSortByIndex has no way of knowing the order of "video" if the workspaces are declared with withScreens. However, getSortByTag works. 
+marshallSort' :: ScreenId -> ([WindowSpace] -> [WindowSpace]) -> ([WindowSpace] -> [WindowSpace])
+marshallSort' s vSort = pScreens . vSort . vScreens where
+        onScreen ws = unmarshallS (tag ws) == s
+        vScreens    = filter onScreen
+        pScreens    = id
+
+-- do something with the Nth workspace
+-- can be used with >9 workspaces
+-- to let withNthWorkspace in XMonad.Actions.DynamicWorkspaces be screen aware
+withNthWorkspace' :: (String -> WindowSet -> WindowSet) -> Int -> X ()
+withNthWorkspace' job wnum = do 
+                                sort <- getSortByIndex
+                                ws <- gets (sort . XMonad.StackSet.workspaces . windowset)
+                                sc <- gets (screen . current . windowset)
+                                case test wnum ws sc of
+                                     (w:_) -> windows $ job w
+                                     [] -> return ()
+                                     where test wnum ws sc = drop wnum [ tag j | j<-ws, wsConvert j == sc]
+
+-- screen aware wrapper for copy from XMonad.Actions.CopyWindow
+copyWinTo :: Int->X ()
+copyWinTo j = do
+            sort <- getSortByIndex
+            ws <- gets (sort . XMonad.StackSet.workspaces . windowset)
+            sc <- gets (screen . current . windowset)
+            let wsID = [ tag m | m<-ws, wsConvert m == sc]
+            windows $ copy $ wsID !! (j-1)
+
+-- return current and visible workspaces on a specific monitor
+lookupWorkspace' :: (Eq s)=> s -> StackSet i l a s sd-> [i]
+lookupWorkspace' sc w = [ tag j | Screen j s _ <- current w : visible w , s == sc ] 
+
+-- take a list of Strings such as ["1","2","3","4"] -> 1234
+fromDigits :: [Char] -> Int
+fromDigits = foldl addDigit 0
+   where addDigit num d = 10*num + ord d - 48 
+
+-- return a list of WorkspaceId that denotes the workspaces which contain the focused window copy
+-- for XMonad.Actions.CopyWindow
+wsContainingCopies' :: X [WorkspaceId]
+wsContainingCopies' = do
+    ws <- gets windowset
+    sc <- gets (screen . current . windowset)
+    return $ copiesOfOn (peek ws) (taggedWindows $ workhidden ws sc )
+          where workhidden ws sc = [ Workspace i l a | Workspace i l a <-hidden ws , show' sc == unmarshall' (show i) ]
+
+-- both taggedWIndows and copiesOfOn are not changed, but these two functions are not exported by their original module, so added here for wsContainingCopies'
+-- | Get a list of tuples (tag, [Window]) for each workspace.
+taggedWindows :: [XMonad.StackSet.Workspace i l a] -> [(i, [a])]
+taggedWindows = map $ XMonad.StackSet.tag &&& XMonad.StackSet.integrate' . XMonad.StackSet.stack
```

Comment:
Perhaps you could import the StackSet qualified via `import qualified XMonad.StackSet as W`; this would be more in-tune with the rest of the code base.

Reply:
Yes, I should. The original code use import XMonad.StackSet, and I was hesitated to use either import ... as W or just leave the original version.

Reply:
Yeah the original code was written at a time when this was probably not that common.  I think it's totally fine to change it at this point for more ergonomic use of the StackSet